### PR TITLE
Rework translations handling

### PR DESCRIPTION
**Pull request by pjedrzejewski**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | #2491 #2447 
| License       | MIT

Hey folks!

This PR decouples ResourceBundle from TranslationBundle, keeping them standalone.

Basically, TranslationBundle becomes an extension of ResourceBundle, which you can install if you want to translate your resources.

Also, it cleans up the configuration a bit, before:

``` yaml
sylius_product:
    driver: doctrine/orm
    classes:
        product:
            model: Your\Product
            controller: Your\ProductController
            repository: Your\ProductRepository
            form:
                default: Your\ProductType
            translatable:
                targetEntity: Your\ProductTranslation
                translatable_fields: [name, banana]
        product_translation:
            model: Your\ProductTranslation
            controller: ~
            repository: ~
            form:
                default: Your\ProductTranslationType
```

After:

``` yaml
sylius_product:
    driver: doctrine/orm
    classes:
        product:
            model: Your\Product
            controller: Your\ProductController
            repository: Your\ProductRepository
            form:
                default: Your\ProductType
            translation:
                model: Your\ProductTranslation
                controller: ~
                repository: ~
                form:
                    default: Your\ProductTranslationType
                mapping:
                    fields: [name, banana]
```

SyliusResourceBundle will automatically map this to 2 resources: ``sylius.product`` and ``sylius.product_translation``. Translation has all the perks as normal resource.
If you do not have SyliusTranslationBundle registered, it will simply ignore this configuration. The circular dependency has been removed:

SyliusResourceBundle requires Resource
Translation requires Resource
SyliusTranslationBundle requires SyliusResourceBundle and Translation

I have also dropped the dependency on ``locale.context``, now TranslationBundle has its own LocaleProvider, which can have different implementations:

In Translation component - Return hardcoded locale
In TranslationBundle - Return locale from current request
In CoreBundle - Get locale from ``locale.context``

Things left to do:

* [x] Make all scenarios green
* [x] Fix few specs
* [x] Allow same thing in SyliusResourceBundle ``resources`` configuration
* [x] Rename CurrentLocaleProviderInterface to LocaleProviderInterface
* [x] Update MongoDB driver

Feedback appreciated!

### CODE REVIEW

**Comment by stloyd** [comment]
:+1:

**Comment by amenophis** [comment]
Good news to see the ResourceBundle work standalone !
Will close my PR: https://github.com/Sylius/Sylius/pull/2455

File: src/Sylius/Bundle/ResourceBundle/spec/DependencyInjection/Driver/DoctrineORMDriverSpec.php [hunk 1]
```diff
@@ -71,7 +71,7 @@ function it_should_create_definition_and_get_repository_in_container(ContainerBu
         $container->getParameter("prefix.repository.resource.class")
             ->shouldBeCalled();
 
-        $container->setDefinition(
+        $container->setDefinitin(
```
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/spec/DependencyInjection/Driver/DoctrineORMDriverSpec.php]: why ?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/spec/DependencyInjection/Driver/DoctrineORMDriverSpec.php]: Mistake. :)

File: src/Sylius/Bundle/TranslationBundle/DependencyInjection/Compiler/TranslationListenerPass.php [hunk 2]
```diff
@@ -27,10 +27,23 @@ class TranslationListenerPass implements CompilerPassInterface
...
+            case SyliusResourceBundle::DRIVER_DOCTRINE_ORM:
+                $reference = 'sylius.orm_translatable_listener';
+            break;
+
+            case SyliusResourceBundle::DRIVER_DOCTRINE_MONGODB_ODM:
+                $reference = 'sylius.mongodb_odm_translatable_listener';
+            break;
+        }
+
+        if (null === $reference) {
```
Comment (stloyd) [inline src/Sylius/Bundle/TranslationBundle/DependencyInjection/Compiler/TranslationListenerPass.php]: Why not in switch?

File: src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Translation\Provider;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class CurrentLocaleProvider implements CurrentLocaleProviderInterface
```
Comment (stloyd) [inline src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php]: That `current` sounds a bit strange IMO.

Comment (pjedrzejewski) [inline src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php]: See the todo list. :)

File: src/Sylius/Bundle/TranslationBundle/DependencyInjection/Compiler/TranslationListenerPass.php (hunk 2 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/TranslationBundle/DependencyInjection/Compiler/TranslationListenerPass.php]: Good catch.

File: src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php (hunk 3 above)
Comment (aRn0D) [inline src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php]: The doto list ?

Comment (pjedrzejewski) [inline src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php]: In PR description, I said:

Things left to do:

Make all scenarios green
Fix few specs
Allow same thing in SyliusResourceBundle resources configuration
Rename CurrentLocaleProviderInterface to LocaleProviderInterface
Update MongoDB driver

Comment (aRn0D) [inline src/Sylius/Component/Translation/Provider/CurrentLocaleProvider.php]: My bad... I thought you created a global todo list for sylius....

**Comment by gperdomor** [comment]
with this we can install the product bundle without translations? because i use that bundle but i not need translations, at least for now

**Comment by pjedrzejewski** [comment]
@gonzalovilaseca Hey man, do you have any idea why slug generation would stop working? Anything to do with our custom slug handler?

**Comment by gonzalovilaseca** [comment]
@pjedrzejewski Probably yes, I will have a look.

File: src/Sylius/Bundle/TranslationBundle/Resources/config/services.xml [hunk 4]
```diff
@@ -17,27 +17,31 @@
...
         <parameter key="sylius.translatable.listener.locale.class">Sylius\Bundle\TranslationBundle\EventListener\LocaleListener</parameter>
+        <parameter key="sylius.translation.locale_provider.request.class">Sylius\Bundle\TranslationBundle\Provider\RequestLocaleProvider</parameter>
     </parameters>
 
     <services>
-        <service id="sylius.translatable.listener" class="%sylius.translatable.listener.class%">
+        <service id="sylius.orm_translatable_listener" class="%sylius.orm_translatable_listener.class%">
             <argument>%sylius.translation.mapping%</argument>
-            <argument>%sylius.locale%</argument>
+            <argument>en_US</argument>
```
Comment (gonzalovilaseca) [inline src/Sylius/Bundle/TranslationBundle/Resources/config/services.xml]: Why hardcode locales in services?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/TranslationBundle/Resources/config/services.xml]: This was temporary, I will put it in configuration.

**Comment by gonzalovilaseca** [comment]
@pjedrzejewski What exactly has stoped working? I downloaded your branch, loaded migrations and all taxon and product slugs seem fine, how can I reproduce the problem?

The only thing I see is that the translation tables have ``es`` as locale, and it should be ``es_ES``

**Comment by aRn0D** [comment]
@pjedrzejewski when do you plan to finish it ?

**Comment by aRn0D** [comment]
@pjedrzejewski We need that (I really need that)!

**Comment by aRn0D** [comment]
Event if you install the translation bundle you still have errors ....  

[Symfony\Component\DependencyInjection\Exception\ParameterNotFoundException]                                                                                                    
  The service "sylius.translatable.listener" has a dependency on a non-existent parameter "sylius.translation.mapping". Did you mean this: "sylius.translation.default.mapping"?

**Comment by pjedrzejewski** [comment]
I'm back at this PR today, wish me good luck. :)

**Comment by aRn0D** [comment]
@pjedrzejewski Great !

**Comment by gperdomor** [comment]
@pjedrzejewski can you provide a ETA for that? like @aRn0D says, We need that, and i have the same errors after install sylius/resource-bundle and his dependencies (sylius/storage, sylius/resource, sylius/translation, sylius/translation-bundle)...

[Symfony\Component\DependencyInjection\Exception\ParameterNotFoundException]

The service "sylius.translatable.listener" has a dependency on a non-existent parameter "sylius.translation.mapping". Did you mean this: "sylius.translation.default.mapping"?

**Comment by amenophis** [comment]
@pjedrzejewski agree with @aRn0D and @gperdomor

**Comment by aRn0D** [comment]
@pjedrzejewski Still working on it ?

**Comment by gperdomor** [comment]
@aRn0D i think so :(... @pjedrzejewski i downgraded to v0.13.0 because I could not make it work TaxonomyBundle like i said in #2589... I hope this solves this issue too :)

**Comment by Shine-neko** [comment]
ping @pjedrzejewski

**Comment by amenophis** [comment]
@pjedrzejewski what's going on this PR ??

**Comment by amenophis** [comment]
Great work !

File: src/Sylius/Bundle/ResourceBundle/DependencyInjection/AbstractResourceExtension.php [hunk 5]
```diff
@@ -261,6 +270,32 @@ protected function loadDatabaseDriver(array $config, LoaderInterface $loader, Co
...
+     * @param ContainerBuilder $container
+     */
+    protected function configureTranslations(array $config, ContainerBuilder $container)
+    {
+        $driver = $config['driver'];
+        $manager = isset($config['object_manager']) ? $config['object_manager'] : 'default';
+        $mapper = new Mapper();
+
+        foreach ($config['classes'] as $model => $classes) {
+            if (array_key_exists('model', $classes) && array_key_exists('translation', $classes)) {
```
Comment (stloyd) [inline src/Sylius/Bundle/ResourceBundle/DependencyInjection/AbstractResourceExtension.php]: `isset($classes['model'], $classes['translation'])`

File: src/Sylius/Bundle/ResourceBundle/DependencyInjection/AbstractResourceExtension.php [hunk 6]
```diff
@@ -261,6 +270,32 @@ protected function loadDatabaseDriver(array $config, LoaderInterface $loader, Co
     }
 
     /**
+     * @param array $config
+     * @param ContainerBuilder $container
+     */
+    protected function configureTranslations(array $config, ContainerBuilder $container)
```
Comment (stloyd) [inline src/Sylius/Bundle/ResourceBundle/DependencyInjection/AbstractResourceExtension.php]: Why this is in resource bundle extension and not i.e. in compiler pass?

**Comment by aRn0D** [comment]
@pjedrzejewski :v:

**Comment by pjedrzejewski** [comment]
I'm running the subtree split now and will test resource bundle standalone, then with translations.

**Comment by pjedrzejewski** [comment]
I'm running the subtree split now and will test resource bundle standalone, then with translations.

**Comment by cdaguerre** [comment]
:+1: Yeeeeah @pjedrzejewski, 0.14 here we come?

**Comment by pjedrzejewski** [comment]
@cdaguerre Yes, blog post in works. :) I will also try to merge all pending bug-fixes etc. But Monday should be the day for v0.14.0. :)

**Comment by pjedrzejewski** [comment]
https://github.com/pjedrzejewski/Sylius/commit/33efb64b41dba4f8b0913154617789077ecebcaf#diff-801545b24fd348917dea21bf86f59f0aL71 This took me so long. For future reference, hehe. :)

**Comment by Shine-neko** [comment]
@pjedrzejewski  In the end your R it is useless. You did not solve the problem of dependence ...

File: src/Sylius/Bundle/ResourceBundle/DependencyInjection/AbstractResourceExtension.php (hunk 6 above)
Comment (isometriks) [inline src/Sylius/Bundle/ResourceBundle/DependencyInjection/AbstractResourceExtension.php]: Yeah, I wonder if you could do like how Symfony does with PrependExtensionInterface. Have TranslatableExtensionInterface that is in the TranslationBundle and search all the bundles in a compiler pass also in TranslationBundle. Then you can even provide the binary for each configure as well:

```php
interface TranslatableExtensionInterface
{
    const CONFIGURE_TRANSLATION = 32;
}
```

**Comment by gperdomor** [comment]
@pjedrzejewski please update the docs too :)

**Comment by pjedrzejewski** [comment]
@Shine-neko I have not tested it yet, but it should be decoupled. Have you?

**Comment by gperdomor** [comment]
@pjedrzejewski ResourceBundle still require TranslationBundle... On the other hand, is posible disable translations in other sylius bundles? like taxonomy or products?

**Comment by isometriks** [comment]
Can we remove translation-bundle from the resource-bundle composer.json now? Maybe move it to suggest?

**Comment by pjedrzejewski** [comment]
@isometriks OMG... I totally forgot to do that. Should be in suggest.

**Comment by gperdomor** [comment]
@pjedrzejewski composer package is not in sync yet.

**Comment by pjedrzejewski** [comment]
@gperdomor In progress. :) It takes a while, I need to definitely optimize it for this amount of bundles.

**Comment by gperdomor** [comment]
@pjedrzejewski :+1: Yayyy!... One more question, is possible disable translations in other sylius bundles? and in case of not, can you add this feature for future release?

**Pull request merged**